packages/muelu/test/scaling/MMKernelDriver.cpp

```diff
@@ -403,107 +403,122 @@ void Multiply_KokkosKernels(const Xpetra::Matrix<Scalar,LocalOrdinal,GlobalOrdin
+      //      std::cout<< "Kokkos::Compat::KokkosOpenMPWrapperNode::execution_space::concurrency() = "<<Kokkos::Compat::KokkosOpenMPWrapperNode::execution_space::concurrency()<<std::endl;
+
+
+
+      // **********************************
+      // Copy in the data for LTG
+      tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("MM LTG: CopyIn")));
+
+      Tpetra::CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node> Aview, Bview;
+      Aview.origMatrix   = Au;
+      Aview.origRowMap   = Au->getRowMap();
+      Aview.rowMap       = Au->getRowMap();
+      Aview.colMap       = Au->getColMap();
+      Aview.domainMap    = Au->getDomainMap();
+      Aview.importColMap = Teuchos::null;
+
+      Bview.origMatrix   = Bu;
+      Bview.origRowMap   = Bu->getRowMap();
+      Bview.rowMap       = Bu->getRowMap();
+      Bview.colMap       = Bu->getColMap();
+      Bview.domainMap    = Bu->getDomainMap();
+      Bview.importColMap = Teuchos::null;
+
+
+      // Because we're in serial...
+      Cnc->replaceColMap(Bu->getColMap());
+
+      // Bcol2Ccol is trivial
+      lo_view_t Bcol2Ccol(Kokkos::ViewAllocateWithoutInitializing("Bcol2Ccol"),Bview.colMap->getNodeNumElements()), Icol2Ccol;
+      const LO colMapSize = static_cast<LO>(Bview.colMap->getNodeNumElements());
+      Kokkos::parallel_for("Tpetra::mult_A_B_newmatrix::Bcol2Ccol_fill",
+                           Kokkos::RangePolicy<execution_space, LO>(0, colMapSize),
+                           KOKKOS_LAMBDA(const LO i) {
+                             Bcol2Ccol(i) = i;
+                           });
+
+      // Acol2Brow
+      local_map_type Acolmap_local = Aview.colMap->getLocalMap();
+      local_map_type Browmap_local = Bview.origMatrix->getRowMap()->getLocalMap();
+      lo_view_t targetMapToOrigRow(Kokkos::ViewAllocateWithoutInitializing("targetMapToOrigRow"),Aview.colMap->getNodeNumElements());
+      lo_view_t targetMapToImportRow;
+      Kokkos::parallel_for("Tpetra::mult_A_B_newmatrix::construct_tables",range_type(Aview.colMap->getMinLocalIndex(), Aview.colMap->getMaxLocalIndex()+1),KOKKOS_LAMBDA(const LO i) {
+        GO aidx = Acolmap_local.getGlobalElement(i);
+        LO B_LID = Browmap_local.getLocalElement(aidx);
+        if (B_LID != LO_INVALID) {
+          targetMapToOrigRow(i)   = B_LID;
+          //        targetMapToImportRow(i) = LO_INVALID;
+        } else {
+          // This shouldn't happen here
+        }
+      });
+      tm = Teuchos::null;
+      Au->getComm()->barrier();
+
+      // **********************************
+      // Multiply
+      Teuchos::RCP<Teuchos::ParameterList> params;
+      tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("MM LTG: Multiply")));
+      Tpetra::MatrixMatrix::ExtraKernels::mult_A_B_newmatrix_LowThreadGustavsonKernel(Aview,Bview,targetMapToOrigRow,targetMapToImportRow,Bcol2Ccol,Icol2Ccol,*Cnc,Cimport,std::string("LTG Test"),params);
+
+      tm = Teuchos::null;
+      Au->getComm()->barrier();
```

Comment:
Why do you need a barrier here?

Reply:
@mhoemmen You'll have to ask @csiefer2 about that. I didn't actually write this function, I only wrapped in a struct so that it is OpenMP-specialized (so the indent makes it look like I wrote the whole thing).